### PR TITLE
Added decimal field to keystone-next with precision, scale

### PR DESCRIPTION
**Pull request by MurzNN**
Added decimal field type like in #4907, but also extended base class (`packages/fields/src/types/Decimal/Implementation.js` for passing precision, scale values to database column in knex.

In GraphQL query field is represented as string, because of missing "Decimal" type in default GraphQL schema.

There are exists extension https://www.npmjs.com/package/graphql-type-decimal to solve this, maybe add it to keystone-next? Or keep decimal as string, like now?

### CODE REVIEW

File: packages-next/fields/src/types/decimal/index.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+export const decimal = <TGeneratedListTypes extends BaseGeneratedListTypes>(
+  config: DecimalFieldConfig<TGeneratedListTypes> = {}
+): FieldType<TGeneratedListTypes> => ({
+  type: Decimal,
+  config,
+  views: resolveView('decimal/views'),
+  getAdminMeta: () => ({
+    precision: config.precision || null,
+    scale: config.precision || null,
```
Comment (timleslie) [inline packages-next/fields/src/types/decimal/index.ts:27]:
```suggestion
    scale: config.scale || null,
```

File: packages-next/fields/src/types/decimal/index.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
+import { Decimal } from '@keystonejs/fields';
+import type { FieldType, BaseGeneratedListTypes, FieldDefaultValue } from '@keystone-next/types';
+import { resolveView } from '../../resolve-view';
+import type { FieldConfig } from '../../interfaces';
+
+export type Int = number & { __int__: void };
+
```
Comment (timleslie) [inline packages-next/fields/src/types/decimal/index.ts:7]:
```suggestion
```

I'd prefer not to use branded types here. While they send a soft signal to the reader, they make for more verbose code for the user, who has to cast their argument to `Int` wherever they use this. They also don't provide any type safety, as the user can case any value to `Int`, whether it's actually an integer value or not.

File: .changeset/good-melons-whisper.md [hunk 3]
```diff
@@ -0,0 +1,6 @@
+---
+'@keystone-next/fields': minor
+'@keystonejs/fields': minor
+---
+
+Added decimal field type for keystone-next with precision, scale
```
Comment (timleslie) [inline .changeset/good-melons-whisper.md:6]:
```suggestion
Added a `decimal` field type to keystone-next.
```

**Review by timleslie — CHANGES_REQUESTED** [review]
Hi @MurzNN, this is all looking pretty good. A couple of minor changes in comments below.

Could I also ask you to update the docs at `docs-next/pages/apis/fields.mdx` to include the `decimal` type.

Now I look I see we're also missing docs for the `float` type. Do you think you could add some docs for that field type as well as a separate PR?

Thanks! 🙏

**Comment by MurzNN** [comment]
I added the documentation for both (`decimal` and `float`) types here, hope my English is not so bad ;)

File: docs-next/pages/apis/fields.mdx [hunk 4]
```diff
@@ -178,6 +178,84 @@ export default config({
...
+});
+```
+
+### decimal
+
+A `decimal` field represents a decimal value.
+
+Options:
+
+- `defaultValue` (default: `undefined`): Can be either a decimal value or an async function which takes an argument `({ context, originalItem })` and returns an decimal value.
```
Comment (timleslie) [inline docs-next/pages/apis/fields.mdx:224]:
```suggestion
- `defaultValue` (default: `undefined`): Can be either a decimal value or an async function which takes an argument `({ context, originalItem })` and returns a decimal value.
```

File: docs-next/pages/apis/fields.mdx [hunk 5]
```diff
@@ -178,6 +178,84 @@ export default config({
...
+  This value will be used for the field when creating items if no explicit value is set.
+  `context` is a [`KeystoneContext`](./context) object.
+  `originalItem` is an object containing the data passed in to the `create` mutation.
+- `precision` (default: `null` treated as `18`): Maximum number of digits that are present in the number.
+- `scale` (default: `null` treated as `4`): Maximum number of decimal places.
```
Comment (timleslie) [inline docs-next/pages/apis/fields.mdx:229]:
```suggestion
- `precision` (default: `18`): Maximum number of digits that are present in the number.
- `scale` (default: `4`): Maximum number of decimal places.
```

File: packages/fields/src/types/Text/Implementation.js [hunk 6]
```diff
@@ -4,10 +4,11 @@ import { KnexFieldAdapter } from '@keystonejs/adapter-knex';
 import { PrismaFieldAdapter } from '@keystonejs/adapter-prisma';
 
 export class Text extends Implementation {
-  constructor(path, { isMultiline }) {
+  constructor(path, { isMultiline, maxLength }) {
```
Comment (timleslie) [inline packages/fields/src/types/Text/Implementation.js:7]: I think these `Text` changes slipped in from your other PR.

**Review by timleslie — CHANGES_REQUESTED** [review]

File: docs-next/pages/apis/fields.mdx [hunk 7]
```diff
@@ -178,6 +178,84 @@ export default config({
 });
 ```
 
+### float
```
Comment (timleslie) [inline docs-next/pages/apis/fields.mdx:181]: I've pulled these changes into #4961 👍

Comment (MurzNN) [reply docs-next/pages/apis/fields.mdx:181]: Thanks!

File: packages/fields/src/types/Text/Implementation.js (hunk 6 above)
Comment (MurzNN) [reply packages/fields/src/types/Text/Implementation.js]: Sorry for this, I remove those code.

**Comment by timleslie** [comment]
This is looking good to me now, thanks @MurzNN.

@mitchellhamilton Could you please take a look at the frontend code? Thanks!

**Review by mitchellhamilton — CHANGES_REQUESTED** [review]
Can you remove all the parseFloat? parseFloat will lose precision

**Comment by MurzNN** [comment]
> Can you remove all the parseFloat? parseFloat will lose precision

@mitchellhamilton, but `.toFixed()` works only with numbers, and we got the string from input widget, not?
```js
'3.14159'.toFixed(2)
// TypeError: "3.14159".toFixed is not a function
```
I can't find the better way for convert string to fractional number, than `parseFloat`.

**Comment by mitchellhamilton** [comment]
You don't need to convert it to anything, the GraphQL API accepts a string

**Comment by MurzNN** [comment]
> You don't need to convert it to anything, the GraphQL API accepts a string

But we need to round the fractional digits to selected scale, for pass the `3.14159` as `3.142` with scale = 3 in field settings, that done via `toFixed(config.fieldMeta.scale)` function.

Or you talk only about `graphql` function, where we can keep the string untouched?

**Comment by mitchellhamilton** [comment]
> But we need to round the fractional digits to selected scale, for pass the 3.14159 as 3.142 with scale = 3 in field settings, that done via toFixed(config.fieldMeta.scale) function.

That should be done by the API/database, not the frontend.

**Comment by MurzNN** [comment]
Ok, I remove parseFloat.

File: packages-next/fields/src/types/decimal/index.ts [hunk 8]
```diff
@@ -1,3 +1,4 @@
+// @ts-ignore
```
Comment (gautamsi) [inline packages-next/fields/src/types/decimal/index.ts:1]: why add it at the top, you can add it where it is needed.

Comment (mitchellhamilton) [reply packages-next/fields/src/types/decimal/index.ts:1]: It's not to the whole file(`@ts-ignore` doesn't work like that), it's to the import on the next line.

Comment (MurzNN) [reply packages-next/fields/src/types/decimal/index.ts:1]: I simply copy-pasted this trick from `packages-next/fields/src/types/integer/index.ts` file :)

**Review by mitchellhamilton — APPROVED** [review]

**Pull request merged**